invisible_cities/cities/detsim_simulate_electrons.py

```diff
@@ -0,0 +1,140 @@
+import numpy as np
+
+from typing import List
+from typing import Tuple
+
+from invisible_cities.cities.detsim_get_psf import binedges_from_bincenters
+
+#######################################
+######### ELECTRON SIMULATION #########
+#######################################
+def generate_ionization_electrons(wi          : float,
+                                  fano_factor : float,
+                                  energies    : np.ndarray) -> np.ndarray:
+    """ generate ionization secondary electrons from energy deposits
+
+    Parrameters:
+        :wi: float
+            ionization yield
+        :fano_factor: float
+            fano-factor. related with the deviation in ionization electrons
+        :energies: np.ndarray
+            energy hits
+    Returns:
+        :nes: np.ndarray
+            the ionization electrons per hit
+    """
+    nes  = np.array(energies/wi, dtype = int)
+    pois = nes < 10
+    nes[ pois] = np.random.poisson(nes[pois])
+    nes[~pois] = np.round(np.random.normal(nes[~pois], np.sqrt(nes[~pois] * fano_factor)))
+    return nes
+
+
+def drift_electrons(lifetime       : float,
+                    drift_velocity : float,
+                    zs             : np.ndarray,
+                    electrons      : np.ndarray) -> np.array:
+    """ returns number of electrons due to lifetime loses from secondary electrons
+
+    Parameters:
+        :lifetime: float
+            electron lifetime
+        :drif_velocity: float
+            drif velocity at the active volume of the detector
+        :zs:
+            the z hits
+        :electrons:
+    Returns:
+        :nes: np.ndarray
+            the drifted electrons per initial hit
+    """
+    ts  = zs / drift_velocity
+    nes = electrons - np.random.poisson(electrons * (1. - np.exp(-ts/lifetime)))
+    nes[nes < 0] = 0
+    return nes
```

Comment:
My understanding is the following:

a. The exponential behaviour of attachment is n'= n*exp(-t/lt), being n' the ie that arrive at the EL from a hit with n ie.
b. Then you define the psf that a single ie reach the EL, that would be: p(t) = exp(-t/lt)/lt. This would be the time distibution of arrival times of ie.
I don't find any physical motivated reason to define this since the ies are associated with a Z position, then an arrival time.
c. Using the cumulative of this distribution (C(t)), you compute a random arrival time (tr) for a single ie as tr=-lt*log(1-C(t)).
d. If the real arrival time (the one from the hit) is less that this tr, then the ie has reached the EL.

This is mathematically correct but it is difficult to me to understand the physical meaning of defining this distribution.

Where is this code coming from?

Reply:
It is just the correct way to sample an exponential distribution...
https://en.wikipedia.org/wiki/Exponential_distribution#Generating_exponential_variates

`-lifetime * np.log(np.random.uniform(size=n_ie)) ` generates n_ie numbers that are distributed according to 1/lifetime exp(-t/lifetime). Then we compare those numbers with the drift time, which determines whether or not each ie reached the gate, as you say. `np.count_nonzero` simply counts how many of the electrons reached the gate.

> I don't find any physical motivated reason to define this since the ies are associated with a Z position, then an arrival time.

I don't know what do you mean here. drift_time = z / drift_velocity and the attachment is parameterized as a function of that. It could be parameterized as a function of z, but it is more common (and suitable to compare with data) to do it as a function of drift time.

Reply:
Yes, this is the method for generating randoms using the "inverse transform sampling" that I mentioned in 3.

Probably I am not explaning myself good enought. I mean that the line of thought is different than I expected.
Imagine a single hit at the active volume, so the pair of values (n, t=Z/dv), being n the number of ionization electrons and t the drift time. We want to compute the number of those electrons reaching the EL (n') from this hit.

-My thought is: just compute n'= n*exp(-t/lt).

-Your thought is a bit different: you generate n exponentially distributed arrival times (tr), and n' would be the number of those times smaller than the hit time (count(tr>t)).

Both ways are mathematically equivalent somehow, but I found yours a bit more confused since it is based in a time-of-arrrival distribution, but the time of each hit is something fixed for each hit.

Reply:
The exponential trick is accurate enough for a large number of initial ie. For a small number, it is not. Imagine a small  (5, for instance) number of atoms decaying. The exponential distribution will tell you the *average* number of atoms left after some time t. If you want to simulate the actual number, you need to do something like what I propose. However, if you have 10000 atoms and you don't wait too long, the number of atoms remaining will be large enough for individual decays to dilute in the global trend.